aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl

```diff
@@ -156,6 +157,48 @@ storage:
             --net=host \
             --dns=host \
             --exec=/bootkube -- start --asset-dir=/assets "$@"
+    - path: /opt/etcd-rejoin
+      filesystem: root
+      mode: 0555
+      contents:
+        inline: |
+          #!/bin/bash
+          set -eou pipefail
+          # Script to rejoin as fresh node to an existing etcd cluster when the node cannot join.
+          # The script should be added to /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf
+          # as ExecStopPost=-/opt/etcd-rejoin because it should run when the etcd service failed and because it
+          # makes use of the environment variables set there for etcd-member.service.
+          # It only runs when etcd failed to join the cluster (e.g., after repovisioning the node or after
+          # the node was down for too long and got kicked out).
+          # Skip if node was not fully provisioned yet.
+          if [ ! -d "/etc/ssl/etcd/" ]; then exit 0; fi
+          # Service just got stopped
+          if [ "$EXIT_CODE" = "killed" ]; then exit 0; fi
+          export ETCDCTL_API=3
+          urls=$(echo "$ETCD_INITIAL_CLUSTER" | tr "," "\n" | cut -d "=" -f 2 | tr "\n" "," | head -c -1)
+          # Double $$ for terraform
+          endpoints="$${urls//2380/2379}"
+          # Assemble etcdctl arguments
+          ARGS="--cacert=/etc/ssl/etcd/etcd-client-ca.crt --cert=/etc/ssl/etcd/etcd-client.crt --key=/etc/ssl/etcd/etcd-client.key --endpoints=$endpoints"
+          # Check if the node is unhealthy and should rejoin
+          unhealty=$((etcdctl endpoint health $ARGS 2> /dev/stdout | grep "is unhealthy" | grep "$ETCD_NAME") || true)
+          if [ -z "$unhealty" ]; then exit 0; fi
+          # Find and remove old member ID if still exists
+          ID=$((etcdctl member list $ARGS | grep "$ETCD_NAME" | cut -d "," -f 1) || true)
+          if [ ! -z "$ID" ]; then
+            etcdctl member remove "$ID" $ARGS
+          fi
+          # Re-add this node as new member
+          etcdctl member add "$ETCD_NAME" --peer-urls="$ETCD_INITIAL_ADVERTISE_PEER_URLS" $ARGS
+          # Let this node join an existing cluster as fresh member (therefore, without state)
+          rm -rf /var/lib/etcd
+          if [ -z "$(grep ETCD_INITIAL_CLUSTER_STATE=existing /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf)" ]; then
```

Comment:
I don't think we have a better way to reference the systemd unit other than hardcoding, right?

Reply:
It's all in the controller.yaml.tmpl file and if the name is changed there in one line I expect that people will search the whole file for that name. Worst case is that it creates a new drop-in file with that line, which would also work.

Reply:
Oh, makes total sense. Sorry! :)